Add copy to clipboard support for install command

sidharthv96 opened the pull request:
Related to https://github.com/Homebrew/formulae.brew.sh/pull/330

MikeMcQuaid commented on assets/css/styles.scss:
```diff
@@ -210,6 +210,27 @@ div .highlight {
 
+.inline {
+  display: inline-flex;
+	@include border-radius(0.4em);
```
This indentation seems pretty weird?

sidharthv96 replied:
![image](https://user-images.githubusercontent.com/10703445/94565959-731c3100-0287-11eb-9046-656e7ab4cd86.png)

Wonder why, looks fine locally.

sidharthv96 replied:
Will just reformat it again.

sidharthv96 replied:
![image](https://user-images.githubusercontent.com/10703445/94566194-b37baf00-0287-11eb-92c5-fb5a40fb4960.png)

 Haha, tabs vs spaces 😂

sidharthv96 replied:
Fixed.

MikeMcQuaid commented on _includes/install_command.html:
```diff
@@ -0,0 +1,24 @@
+        navigator.clipboard.writeText(`{{include.code}}`);
+    }
+
+    if (navigator.clipboard) {
+        document
+            .getElementById("copyButton")
+            .style.removeProperty("display");
+    }
```
Sorry to keep changing my mind on this but: I think it'd be fine to use that library you originally proposed for this if it made this easier: how about this code instead:

- is in a seperate `.js` file
- if `navigator.clipboard` exists: looks for all elements with class (not ID) `copyable` 
- writes a clipboard button to the right of those elements
- adds the logic to copy the contents of the current element's text (rather than passing it through and duplicating it in Jekyll)

sidharthv96 replied:
Yeah, I've done something similar.

Don't think we need the other library.

sidharthv96 commented:
![image](https://user-images.githubusercontent.com/10703445/94565661-2173a680-0287-11eb-8aac-75f12db343e9.png)


The copyable idea can be applied to other installs also...

sidharthv96 commented:
How should I add a js that can be shared across both sites?

MikeMcQuaid commented:
> Just add in assets and link it?

Yup!

sidharthv96 commented:
How about in formulae ? Cross link it to the brew.sh js?

sidharthv96 commented:
Applied copyable to suitable commands in homepage.

![image](https://user-images.githubusercontent.com/10703445/94571852-02c4de00-028e-11eb-8170-90b1017df696.png)

MikeMcQuaid commented:
> Applied copyable to suitable commands in homepage.

Let's leave it just for the installation line at the top 👍🏻

sidharthv96 commented:
> Let's leave it just for the installation line at the top 👍🏻

How about atleast to `brew install wget` and  `brew cask install firefox` too?

It works perfectly, just wanna enjoy the fruits of the generic code. 😅

sidharthv96 commented:
Any specific reason for not wanting it for the other commands?

MikeMcQuaid reviewed (commented):
> Any specific reason for not wanting it for the other commands?

I don't see people copy-pasting them. These are things people read before installation so they don't work (yet).

MikeMcQuaid commented on _layouts/index.html:
```diff
@@ -8,8 +8,11 @@
       <div class="group row">
         <h2 id="install">{{ t.pagecontent.install.install }}</h2>
         <br>
-        <pre style='clear:both;text-align:center;margin-bottom:0.9em'><code id='selectable' onclick="selectText(this)">/bin/bash -c &quot;$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)&quot;</code></pre>
```
Might be nice to keep the `selectText` bit?

sidharthv96 replied:
Done.

sidharthv96 commented on _layouts/base.html:
```diff
@@ -151,6 +151,7 @@ <h1><a href="{{ site.baseurl }}/">{{ site.title }}</a></h1>
         ));
       };
     </script>
+    <script src="{{ "/assets/js/copy.js" | relative_url }}" async></script>
```
@MikeMcQuaid will this load properly in formulae ?

Otherwise I could add the js to the script tag above, and avoid a separate js request.

MikeMcQuaid replied:
> Otherwise I could add the js to the script tag above, and avoid a separate js request.

Yeh, good idea 👍🏻. You'll want to make the JS a function called by something like `<body onload="setupCopyables()">`

sidharthv96 replied:
Done!

sidharthv96 commented:
> I don't see people copy-pasting them. These are things people read before installation so they don't work (yet).

Once they run the initial install command, they could come back and try the install right?
It's easier for them to have a one click command, and doesn't hurt to have it anyway :)

SMillerDev commented:
> Once they run the initial install command, they could come back and try the install right?
> It's easier for them to have a one click command, and doesn't hurt to have it anyway :)

I think it makes it sorta confusing to read, you're reading the docs but it's trying to get you to install wget and firefox. What if I don't even want those? What if I'm on Linux and casks don't work for me? I'd say the cleanest solution is to stick with the parts that people definitely want to copy.

MikeMcQuaid commented:
Tested this locally, this is looking great! A few needed tweaks:

The "selection on click" now also selects the 📋 and goes "deeper" than it did before:
<img width="824" alt="Screenshot 2020-10-01 at 09 13 18" src="https://user-images.githubusercontent.com/125011/94784682-6eaa6200-03c6-11eb-9dc8-9105ffb8cdbd.png">

If you zoom in the alignment for the separate section isn't quite right:
<img width="170" alt="Screenshot 2020-10-01 at 09 11 47" src="https://user-images.githubusercontent.com/125011/94784748-84b82280-03c6-11eb-931f-1c94db05dcc8.png">

It might be nice to have a hover/click visual difference in the icon (even just in the background, maybe)? Check out what GitHub does:

<img width="88" alt="Screenshot 2020-10-01 at 09 12 30" src="https://user-images.githubusercontent.com/125011/94784841-a3b6b480-03c6-11eb-809e-1b268f42fc1a.png">
<img width="79" alt="Screenshot 2020-10-01 at 09 12 32" src="https://user-images.githubusercontent.com/125011/94784843-a44f4b00-03c6-11eb-8d70-df90b99cfe5f.png">
<img width="88" alt="Screenshot 2020-10-01 at 09 12 36" src="https://user-images.githubusercontent.com/125011/94784844-a4e7e180-03c6-11eb-924f-a6f572738096.png">

sidharthv96 commented:
![image](https://user-images.githubusercontent.com/10703445/94787387-dfb23f80-03ef-11eb-97ac-ae8c985db481.png)

On Hover
![image](https://user-images.githubusercontent.com/10703445/94787709-47688a80-03f0-11eb-9a49-f80423a3d49e.png)


Should the Icon be SVG? Do we have fa icons available ?

MikeMcQuaid commented:
Hover looks good! No SVG icons available. On click could change back to the non-hover colour to indicate we've taken an action.

sidharthv96 commented:
There is a chrome rendering issue (bug?) here.

![image](https://user-images.githubusercontent.com/10703445/94789511-a0392280-03f2-11eb-862a-8db7985a2946.png)
![image](https://user-images.githubusercontent.com/10703445/94789172-3587e700-03f2-11eb-9479-e619714f2109.png)

--- 
Firefox looks fine
![image](https://user-images.githubusercontent.com/10703445/94789413-826bbd80-03f2-11eb-9c6b-17d20916c7fd.png)
![image](https://user-images.githubusercontent.com/10703445/94789470-93b4ca00-03f2-11eb-9d4f-3f6c8713a0ba.png)

sidharthv96 commented:
![image](https://user-images.githubusercontent.com/10703445/94790172-8ea44a80-03f3-11eb-82f3-4ceabf3b8783.png)

Looks like a chromium bug, it's visible inside VSCode also.

sidharthv96 commented:
![image](https://user-images.githubusercontent.com/10703445/94790690-4174a880-03f4-11eb-9fc9-e756be4e0ea8.png)

Will revert back after 1s.

sidharthv96 commented:
Adding 2 SVG icons might look more cleaner than using unicode?

MikeMcQuaid commented:
> Will revert back after 1s.

Good idea!

> Adding 2 SVG icons might look more cleaner than using unicode?

I think the unicode "feels" more like Homebrew.

sidharthv96 commented:
So I guess we can wrap it up then?

MikeMcQuaid commented:
Looking great! Almost there! A few tiny things:

<img width="136" alt="Screenshot 2020-10-01 at 16 54 39" src="https://user-images.githubusercontent.com/125011/94833353-ee582100-0406-11eb-8af6-a22877b06a07.png">

Still some very slight alignment issues here (on Safari).

<img width="738" alt="Screenshot 2020-10-01 at 16 54 59" src="https://user-images.githubusercontent.com/125011/94833374-f912b600-0406-11eb-989d-81c88ee4d615.png">
<img width="923" alt="Screenshot 2020-10-01 at 16 54 56" src="https://user-images.githubusercontent.com/125011/94833377-f9ab4c80-0406-11eb-8524-c1d041502738.png">

The `selectText` selection area seems to have changed. Not a big deal but might be nice to address.

Relatedly, I wonder whether the `selectText` stuff can go away entirely? Given it won't work if you have JS disabled and the clipboard copy is a much nicer version if you don't. Thoughts?

sidharthv96 commented:
> Still some very slight alignment issues here (on Safari).

Safari was adding extra margin for buttons. Fixed.
![image](https://user-images.githubusercontent.com/10703445/94839272-60932a80-0434-11eb-9487-e6182b63c4e1.png)


> The selectText selection area seems to have changed. Not a big deal but might be nice to address.

That's because we are using the highlight instead of a plain code block, so there is a pre with a bottom margin. If I select the inner code, it can be back to normal. `elem = elem.getElementsByTagName("code")[0];`
![image](https://user-images.githubusercontent.com/10703445/94838774-b3201700-0433-11eb-9494-a312bbbcb4e6.png)

> Relatedly, I wonder whether the selectText stuff can go away entirely? Given it won't work if you have JS disabled and the clipboard copy is a much nicer version if you don't. Thoughts?

Haha, that's why I had removed it at first. I'm fine with either approach, but I think it is a bit distracting to have the copy button copy to clipboard and also select the text, the user might get confused and hit copy again. Also, the select text is broken in chrome for me.

sidharthv96 commented:
The emoji spacing is indeed a chrome issue in macOS.
https://bugs.chromium.org/p/chromium/issues/detail?id=596223

MikeMcQuaid commented:
Thanks so much for your amazing first contribution and patience! Without people like you submitting PRs we couldn't run this project. You rock, @sidharthv96!

sidharthv96 commented:
Thanks @MikeMcQuaid !
Really appreciate the time and effort you put into making this so much better than the initial draft. 😄

MikeMcQuaid commented:
You're welcome, thanks for your patience with all the rounds of review!
